proposals/0051-options-local-pragma.rst

```diff
@@ -0,0 +1,145 @@
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/0>`_.
+.. sectnum::
+.. contents::
+
+We propose to add functionality for switching GHC-options locally. For instance, this code
+::
+ {-# OPTIONS_GHC -Wno-name-shadowing #-}
+ 
+ f = {-# OPTIONS_LOCAL -Wname-shadowing #-}
+   let a = () in
+   let a = 5 in
+     a + a
+will result in a  ``-Wname-shadowing`` warning at line
+::
+ let a = 5 in
+
+This proposal rises some ticket problems:
+ 1. https://gitlab.haskell.org/ghc/ghc/issues/602
+ 2. https://gitlab.haskell.org/ghc/ghc/issues/10150
+
+Motivation
+------------
+
+There are the following reasons to implement this pragma.
+
+Local control of warnings
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To allow control of warnings when you are confident in your decisions without missing really important warnings.
+
+Warnings are compiler recomendations and often they can be really useful. They allow us to avoid stupid mistakes and point to *weak* places in the code. You should pay attention to them. But sometimes you can be confident in your local code-decision and get some useless warning. It clutters the warnings list and you can miss a really important warning.
+
+Consider two situations:
+ 
+1. You want to enable a warning locally (while it will remain disabled in the rest of the file). Consider an example `suppress orphan instance warning per instance <https://gitlab.haskell.org/ghc/ghc/issues/10150>`_.
+   ::
+    module Foo (
+      ApplyFunc(..)
+    ) where
+
+    class ApplyFunc f where
+      func :: (a -> b) -> f a -> f b
+
+   
+    module Bar (
+      Box(..)
+    ) where
+
+    data Box a = Empty
+               | Content a 
+
+   
+    {-# OPTIONS -Worphans #-}
+    module Baz where
+
+    import Foo
+    import Bar
+
+    instance ApplyFunc Box where
+      func f Empty       = Empty
+      func f (Content a) = Content $ f a
+
+2. You want to disable a warning locally (while it will remain enabled in the rest of the file):
+   ::
+    {-# OPTIONS_GHC -Wunused-do-bind #-}
+
+    main :: IO ()
+    main = {-# OPTIONS_LOCAL -Wno-unused-do-bind #-} do
+      getLine
+      return ()
+
+`Ticket 602 <https://gitlab.haskell.org/ghc/ghc/issues/602>`_ links with this motivation point.
+
+Using language extensions locally
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Using language extensions locally to avoid using it in other places by mistake.
+
+Language extensions allow to add useful things but "with great power there must also come great responsibility". As a result, sometimes it leads to a lot of mistakes and complicates code.
+
+You can enable a language extension locally (and it will remain disabled in the rest of the file).
+::
+ {-# OPTIONS_GHC -Wno-name-shadowing #-}
+
+ x' = {-# OPTIONS_LOCAL -XPartialTypeSignatures #-}
+   let a :: _
+       a = ()
+    in ()
+
+`Comment from the ticket 602 <https://gitlab.haskell.org/ghc/ghc/issues/602#note_108677>`_
```

Comment:
Perhaps insert that comment as a proper quote into the proposal so that the reader doesn't have to follow the link.

Same goes for other links to tickets and comments.